game/players.py

```diff
@@ -212,14 +219,109 @@ def update_agent(self, optimizer) -> None:
+                  store_metadata: bool = True) -> None:
+        """Save learnable parameters, optimizer dictionary and metadata.
+
+        Save data relevant for inference and resuming training with
+        torch.save(). Overrides Player.save_agent().
+        A dictionary is stored at "folder/file_name.pt". It contains:
+            - 'model_state_dict': The state_dict of the model, containing
+                current values of learnable parameters.
+            - 'optim_state_dict': The state_dict of the optimizer.
+                (needed for further training)
+                Note: This is left out if argument on_quit is True.
+            - 'loss_sum': the current loss, ie. self.stats['loss_sum'].
+                (needed for further training)
+        The model_state_dict can be loaded with self.load_network_weights().
+
+        If store_metadata is True, then also stores metadata as .json at
+        file_name_metadata.json. Included metadata is
+            - name of the agent class
+            - name of the script invoked to call this function
+            - current hash of the git repository the script resides in
+            - id of the neptune run used for logging, empty string if not used
+            - name of the optimizer class used, "NoneType" if data is stored
+                upon quitting the environment.
+
+        WARNING: The function overwrites existing files without warning.
+
+        Args:
+            file_name (str): name of the file to be stored. If file extension
+                is not specified, default is ".pt". Both ".pt" and ".pth" are
+                accepted. The metadata file will append "metadata.json" to
+                the provided filename (without extension).
+            optimizer (torch.optim.Optimizer): Current optimizer object.
+            on_quit (bool, optional): Is the function called at quit-event by
+                Env.check_user_events()? Defaults to False.
+            folder (str, optional): target directory for parameters and
+                metadata, can be "" to store in the same folder as the script
+                calling the function. Defaults to "learned_weights".
+            store_metadata (bool, optional): Also save. Defaults to True.
+
+        Raises:
+            ValueError: When user provides a file_name with invalid extension
+        """
+        name, ext = path.splitext(file_name)
+        if ext == '':
+            file_name += '.pt'
+        elif ext not in ('.pt', '.pth'):
+            raise ValueError(
+                'Invalid file extension, must be .pt, .pth or not specified.'
+                )
+
+        # ensure that folder exists prior to saving (torch needs this)
+        if not path.isdir(folder) and folder != '':
+            mkdir(folder)
+        
+        # save learnable parameters
+        if not folder.endswith('/') and folder != '':
+            folder += '/'
+        full_path = folder + file_name
+        if not on_quit:
```

Comment:
Hvorfor er det at optim_state_dict ikke bliver gemt hvis man har quitet? og kun hvis man ikke quiter?

Reply:
Fordi optimizeren ikke er direkte tilgængelig for hverken env eller agenten, men kun eksisterer ude i træningsscriptet. Hvis vi vil gemme optimizeren ved quit er vi således nødt til at give env optimizeren som input, så den kan give det som argument til `save_agent()` i `check_user_events()`. Giver det mening? Ellers må vi lige tage det mundtligt mellem jul og nytår.

Så jeg kan godt ændre det, men det kræver at ændre lidt i env

Reply:
Tænker ikke at der skal ændret for meget i env, da det kan blive til et større projekt med alt andet, men tænkte mere på om at i stedet for at have det der med input af "on_quit", så kunne sige at optimizer var default "None" og så tjekkede man om den var noget andet end "None". Hvis den er, så gemmer man den ellers så gemmer man den ikke.

Kunne det ikke også løse det?

Reply:
Jo, den tilgang ville også virke. Ville du foretrække at have default behaviour fremfor det ekstra eksplicitte on_quit? Jeg kan godt ændre det, hvis det er. Så kommer vi bare til at have default af en anden type end optimizer egentlig skal være, men det er måske alligevel nemmere at forstå på den måde?

Reply:
Er implementeret nu.